text/0005-coretime-interface.md

```diff
@@ -0,0 +1,183 @@
+
+## Explanation
+
+The interface has two sections: The messages which the Relay-chain is able to receive from the allocating parachain (the *UMP message types*), and messages which the Relay-chain is able to send to the allocating parachain (the *DMP message types*). These messages are expected to be able to be implemented in a well-known pallet and called with the XCM `Transact` instruction.
+
+Future work may include these messages being introduced into the XCM standard.
+
+### UMP Message Types
+
+#### `request_core_count`
+
+Prototype:
+
+```
+fn request_core_count(
+    count: u16,
+)
+```
+
+Requests the Relay-chain to alter the number of schedulable cores to `count`. Under normal operation, the Relay-chain SHOULD send a `notify_core_count(count)` message back.
+
+#### `request_revenue_info_at`
+
+Prototype:
+
+```
+fn request_revenue_at(
+    when: BlockNumber,
+)
+```
+
+Requests that the Relay-chain send a `notify_revenue` message back at or soon after Relay-chain block number `when` whose `until` parameter is equal to `when`.
+
+The period in to the past which `when` is allowed to be may be limited; if so the limit should be understood on a channel outside of this proposal. In the case that the request cannot be serviced because `when` is too old a block then a `notify_revenue` message must still be returned, but its `revenue` field may be `None`.
+
+#### `credit_account`
+
+Prototype:
+
+```
+fn credit_account(
+    who: AccountId,
+    amount: Balance,
+)
+```
+
+Instructs the Relay-chain to add the `amount` of DOT to the Instantaneous Coretime Market Credit account of `who`.
+
+It is expected that Instantaneous Coretime Market Credit on the Relay-chain is NOT transferrable and only redeemable when used to assign cores in the Instantaneous Coretime Pool.
+
+#### `assign_core`
+
+Prototype:
+
+```
+type PartsOf57600 = u16;
+enum CoreAssignment {
+    InstantaneousPool,
+    Task(ParaId),
+}
+fn assign_core(
+    core: CoreIndex,
+    begin: BlockNumber,
+    assignment: Vec<(CoreAssignment, PartsOf57600)>,
+    end_hint: Option<BlockNumber>,
+)
+```
+
+Requirements:
+
+```
+assert!(core < core_count);
+assert!(targets.iter().map(|x| x.0).is_sorted());
+assert_eq!(targets.iter().map(|x| x.0).unique().count(), targets.len());
+assert_eq!(targets.iter().map(|x| x.1).sum(), 57600);
+```
+
+Where:
+- `core_count` is assumed to be the sole parameter in the last received `notify_core_count` message.
+
+Instructs the Relay-chain to ensure that the core indexed as `core` is utilised for a number of assignments in specific ratios given by `assignment` starting as soon after `begin` as possible. Core assignments take the form of a `CoreAssignment` value which can either task the core to a `ParaId` value or indicate that the core should be used in the Instantaneous Pool. Each assignment comes with a ratio value, represented as the numerator of the fraction with a denominator of 57,600.
```

Comment:
I don't get how `assignment` should work? How does the assignment and fraction map to scheduling the assignment to the core? 

I mean for things like `[(Task(1), 0.5), (Task(2), 0.5)]` task 1 and task 2 will be running interleaved all the time?

Reply:
> How does the assignment and fraction map to scheduling the assignment to the core?

It means that the relay-chain should allow the task/assignment to use that proportion of the core's resources. It's better to leave it general like this, because this definition also allows tasks to use 1/2 of the core every relay-chain block, when we get support for having multiple candidates per core at a time.

And scheduling tasks to _specific_ relay-chain blocks is actually really bad - this formulation does work nicely with e.g. #3 , which is a probabilistic scheduler

> I mean for things like [(Task(1), 0.5), (Task(2), 0.5)] task 1 and task 2 will be running interleaved all the time?

Yes, or `(Task(1), 0.25), (Task(2), 0.5), (Instantaneous, 0.25)` would mean that:
  * Task 1 gets to use 1/4 of the core's resources for this timeframe (1 block out of every 4 relay chain blocks)
  * Task 2 gets to use 1/2 of the core's resources (2 blocks out of every 4)
  * Instantaneous gets to allocate 1 block every 4 relay chain blocks.

Reply:
> It's better to leave it general like this, because this definition also allows tasks to use 1/2 of the core every relay-chain block, when we get support for having multiple candidates per core at a time.

But how would that look like? Currently the fraction expresses how much blocks a task can build for a given time range. I assume this time range will be `TIMESLICE` and both sides relay chain/brocker chain will be aware of this? 

We will need something to tell the relay chain that a certain task actually needs only 1/2 of a core and not 1/2 of the time range.

Reply:
It's designed to be probabilistic - over a sufficiently long period it would be scheduled 1/2 of the time, and the Relay-chain is expected to be as fair as it can be, but practical constraints preclude the ability to state anything that is "perfectly fair all the time". We don't explicitly specify a timerange; the relay-chain is expected to just maximise the fairness over a minimal time range.

Reply:
> But how would that look like? Currently the fraction expresses how much blocks a task can build for a given time range. I assume this time range will be TIMESLICE and both sides relay chain/brocker chain will be aware of this?

While it would be valid for the relay-chain side to interpret this interface as giving each chain 1/2 of the timeslice, that would be a pretty bad scheduler. Scheduler implementations (like #3) should probably aim for probabilistic interleaving.